vm/systemSmartContracts/liquidStaking.go

```diff
@@ -199,25 +306,256 @@ func (l *liquidStaking) reDelegateRewardsFromPosition(args *vmcommon.ContractCal
+
+		nonce, err := l.createOrAddNFT(attributes.ContractAddress, newCheckpoint, totalToCreate)
+		if err != nil {
+			l.eei.AddReturnMessage(err.Error())
+			return vmcommon.UserError
+		}
+
+		listNonces = append(listNonces, nonce)
+		listValues = append(listValues, totalToCreate)
+	}
+
+	err := l.sendNFTMultiTransfer(args.CallerAddr, listNonces, listValues, args.Arguments)
+	if err != nil {
+		l.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
 	return vmcommon.Ok
 }
 
-func (l *liquidStaking) unDelegateWithPosition(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+func (l *liquidStaking) returnLiquidStaking(
+	args *vmcommon.ContractCallInput,
+	functionToCall string,
+) vmcommon.ReturnCode {
 	returnCode := l.checkArgumentsWhenPositionIsInput(args)
 	if returnCode != vmcommon.Ok {
 		return returnCode
 	}
 
+	for _, esdtTransfer := range args.ESDTTransfers {
+		_, _, returnCode = l.burnAndExecuteFromESDTTransfer(
+			args.CallerAddr,
+			esdtTransfer,
+			functionToCall,
+		)
+		if returnCode != vmcommon.Ok {
+			return returnCode
+		}
+	}
+
 	return vmcommon.Ok
 }
 
-func (l *liquidStaking) returnPosition(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
-	returnCode := l.checkArgumentsWhenPositionIsInput(args)
+func (l *liquidStaking) burnAndExecuteFromESDTTransfer(
+	callerAddr []byte,
+	esdtTransfer *vmcommon.ESDTTransfer,
+	functionToCall string,
+) (*LiquidStakingAttributes, [][]byte, vmcommon.ReturnCode) {
+	attributes, err := l.getAttributesForNonce(esdtTransfer.ESDTTokenNonce)
+	if err != nil {
+		l.eei.AddReturnMessage(err.Error())
+		return nil, nil, vmcommon.UserError
+	}
+
+	err = l.burnNFT(esdtTransfer.ESDTTokenNonce, esdtTransfer.ESDTValue)
+	if err != nil {
+		l.eei.AddReturnMessage(err.Error())
+		return nil, nil, vmcommon.UserError
+	}
+
+	returnData, returnCode := l.executeOnDestinationSC(
+		attributes.ContractAddress,
+		functionToCall,
+		callerAddr,
+		esdtTransfer.ESDTValue,
+		attributes.RewardsCheckpoint,
+	)
 	if returnCode != vmcommon.Ok {
-		return returnCode
+		return nil, nil, returnCode
 	}
 
-	return vmcommon.Ok
+	return attributes, returnData, vmcommon.Ok
+}
+
+func (l *liquidStaking) executeOnDestinationSC(
+	dstSCAddress []byte,
+	functionToCall string,
+	userAddress []byte,
+	valueToSend *big.Int,
+	rewardsCheckPoint uint32,
+) ([][]byte, vmcommon.ReturnCode) {
+	txData := functionToCall + "@" + hex.EncodeToString(userAddress) + "@" + hex.EncodeToString(valueToSend.Bytes())
+	if rewardsCheckPoint > 0 {
+		txData += "@" + hex.EncodeToString(big.NewInt(int64(rewardsCheckPoint)).Bytes())
+	}
+	vmOutput, err := l.eei.ExecuteOnDestContext(dstSCAddress, l.liquidStakingSCAddress, big.NewInt(0), []byte(txData))
+	if err != nil {
+		l.eei.AddReturnMessage(err.Error())
+		return nil, vmcommon.UserError
+	}
+
+	if vmOutput.ReturnCode != vmcommon.Ok {
+		return nil, vmOutput.ReturnCode
+	}
+
+	return vmOutput.ReturnData, vmcommon.Ok
+}
+
+func (l *liquidStaking) createOrAddNFT(
```

Comment:
createOrAddSFT ? Valid for all these xxxNFTxxx functions.

Reply:
done